docs/blog/20240122-savetheclocktower-modern-tree-sitter-part-6.md

```diff
@@ -0,0 +1,190 @@
+
+(What Maurício calls “definitions” is what `symbols-view` calls “symbols,” and what `ctags` calls, well, “tags.” For simplicity I’ll use the term “symbol” just to align with what Pulsar calls it.)
+
+`ctags` works well enough that you might never notice its drawbacks, but it’s got _plenty_ of drawbacks. It reads files from disk, so it can return inaccurate results if you use it on a file that has been modified since its last save. For the same reason, it doesn’t work _at all_ on new files that haven’t yet been saved. And it needs [special configuration](https://github.com/pulsar-edit/symbols-view/blob/master/lib/ctags-config) for each language it supports — meaning that, even after you’ve written a Pulsar grammar for your newly-invented Language X, you won’t get any symbol-based navigation unless you modify the `symbols-view` package itself and tell `ctags` how to find your language’s symbols.
+
+I know you’re probably tired of hearing me say “Tree-sitter would be _great_ for this task!” — but [code navigation systems](https://tree-sitter.github.io/tree-sitter/code-navigation-systems) really are in its wheelhouse. The trees we’re already using to highlight code and do other useful tasks can be queried to supply symbols much more easily than via `ctags`. And many parsers even come bundled with a query file that does the work of identifying the symbols we’re interested in.
+
+You might have noticed how GitHub can nowadays give you an outline-like view of a source code file, listing lines where methods are defined. That’s all happening through Tree-sitter. If GitHub can use it for symbol navigation, so can we.
+
+## Refactoring `symbols-view`
+
+But to make that happen, we need to change how `symbols-view` works. All it knows about is `ctags`! Could we rip all that out and replace it with a Tree-sitter solution? Yes, but in the process we’d be abandoning support for any languages that don’t yet have Tree-sitter parsers.
+
+A better approach would be to know about _both_ strategies and pick the best one on the fly. So let’s figure out exactly what Maurício’s request — “transforming \[symbols\] into a ‘service’” — means.
+
+### A crash course in services
+
+In Pulsar, **services** are how packages talk to one another. Suppose I’ve authored `package-b` and it depends on another package called `package-a` that someone else has written. I could reach into `atom.packages` and grab the reference to `package-a`, but this feels weird for a number of reasons. For one, it incorrectly assumes that `package-a` has already been activated. It may get activated _after_ `package-b` — or else it may _never_ get activated because the user has disabled or uninstalled it.
+
+But even if `package-b` were able to find and consume package `package-a` this way, it’d create a tight coupling between the two. That coupling would break if `package-a` renamed itself, or if were rewritten and changed implementation details that `package-b` was relying on.
```

Comment:
Whoops! I missed the pronoun here. But “were” is correct because we're in the subjunctive mood.

Reply:
I was struggling with this one as it was technically correct but sounded wrong to my ears 1) due to the missing pronoun and 2) there is a slight garden path on "rewritten and changed..." that got me a bit. As said, entirely optional so feel free to dismiss.

Reply:
I've got a compromise that will delight us both!

> That coupling would break if `package-a` renamed itself, or if it changed implementation details that `package-b` was relying on.

Reply:
Sounds better than both I have to agree